proto/xmtpv4/message_api/message_api.proto

```diff
@@ -5,20 +5,32 @@ package xmtp.xmtpv4;
 
 import "google/api/annotations.proto";
 import "identity/associations/signature.proto";
+import "mls/api/v1/mls.proto";
 
 option go_package = "github.com/xmtp/proto/v3/go/xmtpv4/message_api";
 
-// Authenticated data within the MlsPrivateMessage
+// Data visible to the server that has been signed by the client.
+// Included in one of:
+// - MLS PrivateMessage.authenticated_data for messages and commits
+// - MLS KeyPackageTBS.extensions for key packages
+// - Omitted for Welcome messages
 message AuthenticatedData {
   repeated uint64 last_originator_sids = 1;
 }
 
-// Replaces GroupMessageInput V1
-// To rename or not to rename?
+// Data visible to the server that has been signed by the payer. Fields can
+// be set by the client OR the payer, but the final result will be returned
+// to the client after publishing.
 message ClientEnvelope {
-  // TLS serialized MlsMessageIn, which contains MlsPrivateMessage
-  bytes data = 1;
-  bytes sender_hmac = 2;
+  uint32 target_originator = 1;
+  bytes target_topic = 2;
+  oneof payload {
+    xmtp.mls.api.v1.GroupMessageInput group_message = 3;
+    xmtp.mls.api.v1.WelcomeMessageInput welcome_message = 4;
+    xmtp.mls.api.v1.RegisterInstallationRequest register_installation = 5;
+    xmtp.mls.api.v1.UploadKeyPackageRequest upload_key_package = 6;
+    xmtp.mls.api.v1.RevokeInstallationRequest revoke_installation = 7;
```

Comment:
Reusing all of the old API inputs minimizes client changes, and also allows us to port over any of the old server validation logic, if we choose to keep it. We can also use this field to determine what goes onto blockchain vs replication sub-systems. Note the server is not *forced* to do any custom logic for each type here - it can choose to just serialize all of this and store it in the DB as a byte array.

One difference from the old API will be the outputs from the query API's, which are a slightly different shape. The old API had many different output types, which were mostly equivalent to the input types, with a sequence ID and timestamp added. Our `GatewayEnvelope` basically includes the same metadata on a single type.

I'm hoping that on the client, we can reflect these changes as a thin wrapper on the API client.

Reply:
That seems like a sensible strategy. I wasn't the one pushing for having all different types of APIs for reads/writes, and now that I've been living with it for 6 months I'm not sure it adds much value.